sdks/python/apache_beam/runners/direct/evaluation_context.py

```diff
@@ -99,6 +113,19 @@ def finalize_value_and_get_tasks(self, side_input):
       view.has_result = True
       return result
 
+  def update_watermarks_for_transform(self, ptransform, watermark):
+    # Collect tasks that get unblocked as the workflow progresses
+    unblocked_tasks = []
+    for view in self._transform_to_views[ptransform]:
+      unblocked_tasks.extend(self._update_watermarks_for_view(view, watermark))
+    return unblocked_tasks
+
+  def _update_watermarks_for_view(self, view, watermark):
+    unblocked_tasks = []
+    if watermark.input_watermark == WatermarkManager.WATERMARK_POS_INF:
```

Comment:
Would not waiting for `WatermarkManager.WATERMARK_POS_INF` unblock side inputs only when they are completely finished?

Reply:
Yes, this is part one of a refactor that will deliver the streaming side input functionality.  I've synced with Maria, and she has the other parts of the refactor after this one goes in.  Perhaps we should clarify the title of the PR though.

Reply:
Thank you. Sounds good, let's do that.

Reply:
Done